### PR TITLE
Implement backward for torch.symeig

### LINKED ISSUE #6890: [feature request] the derivative for 'eig' is not implemented
It seems that the derivative for 'eig' is not implemented in torch 0.3.1 .post2.
How could I use the derivate for 'eig'?

### PR DESCRIPTION
**Pull request by sethah**
Partially fixes https://github.com/pytorch/pytorch/issues/6890. (backward pass for non-symmetric eigen-decomposition is not implemented in other packages, e.g. autograd, mxnet, tensorflow, presumably because the eigenvalues can be imaginary for the general case, and AFAIK we cannot support complex numbers).

This patch adds a backward function for the symmetric eigen-decomposition function `torch.symeig`. The formula used is taken from [here](http://eprints.maths.ox.ac.uk/1079/1/NA-08-01.pdf). Unit tests are added to verify correctness.

There is still one outstanding issue, which is how to handle the case where the `symeig` is called with `eigenvectors=False`. In this case, the eigenvectors are returned as a zero tensor, but the backward computation for the eigenvalues depends on the eigenvectors. There was a previous attempt to implement this in https://github.com/pytorch/pytorch/pull/2026, where @apaszke mentioned that the `eigenvectors` argument should be overridden so that they are saved for the backwards pass. The forward code is autogenerated, though, and it isn't clear to me how that would be done. I'd appreciate any guidance. For now, there is a unit test that will fail until that issue is resolved.

### CODE REVIEW

File: tools/autograd/templates/Functions.cpp [hunk 1]
```diff
@@ -973,6 +973,36 @@ Tensor svd_backward(const std::vector<torch::autograd::Variable> &grads, const T
...
+    auto gv = grads[1];
+    
+    auto vt = v.t();
+    
+    Tensor result = at::zeros_like(self);
+    if (gv.defined()) {
+        Tensor F = at::zeros_like(self);
+        F.add_(at::unsqueeze(lambda, 0)).sub_(at::unsqueeze(lambda, 1));
+        F.as_strided({n}, {n + 1}).fill_(INFINITY);
+        F.pow_(-1);
```
Comment (vishwakftw) [inline tools/autograd/templates/Functions.cpp]: You could use `reciprocal`.

**Comment by t-vi** [comment]
I think the backward should at least explicitly fail when the eigenvectors are not kept. Then the user will know what to do to fix it.

**Comment by SsnL** [comment]
I agree with @t-vi . Let's just fail when `eigenvector = false` now. 

One can work around this by make `symeig` always compute `eigenvector` and expose a wrapper function to Python. But that is inefficient when users don't want gradients.

We should have a way to specify grad mode on vs. off forward. I'll post an issue on that. But for this PR, let's just make it an error.

**Comment by SsnL** [comment]
Can you make this change and add a test that assert an except is throw when `eigenvector=false` in backward?

**Comment by sethah** [comment]
I added a test to check the backwards call fails. I wasn't sure if this test needs a cuda version, so if it does then I'll be happy to add it.

I'll fix merge conflicts shortly.

**Comment by sethah** [comment]
ping @SsnL

File: tools/autograd/templates/Functions.cpp [hunk 2]
```diff
@@ -1510,6 +1510,41 @@ Tensor svd_backward(const std::vector<torch::autograd::Variable> &grads, const T
...
+    if (!eigenvectors) {
+        throw std::runtime_error(std::string("cannot compute backward without "
+                                             "computing eigenvectors in forward pass"));
+    }
+    
+    Tensor result = at::zeros_like(self);
+    if (gv.defined()) {
+        Tensor F = at::zeros_like(self);
+        F.add_(at::unsqueeze(lambda, 0)).sub_(at::unsqueeze(lambda, 1));
+        F.as_strided({n}, {n + 1}).fill_(INFINITY);
```
Comment (SsnL) [inline tools/autograd/templates/Functions.cpp]: nit: use `diagonal()` instead

File: tools/autograd/templates/Functions.cpp [hunk 3]
```diff
@@ -1510,6 +1510,41 @@ Tensor svd_backward(const std::vector<torch::autograd::Variable> &grads, const T
...
+    
+    if (!eigenvectors) {
+        throw std::runtime_error(std::string("cannot compute backward without "
+                                             "computing eigenvectors in forward pass"));
+    }
+    
+    Tensor result = at::zeros_like(self);
+    if (gv.defined()) {
+        Tensor F = at::zeros_like(self);
+        F.add_(at::unsqueeze(lambda, 0)).sub_(at::unsqueeze(lambda, 1));
```
Comment (SsnL) [inline tools/autograd/templates/Functions.cpp]: nit: the `zeros_like` and `add_` can be more efficiently written as `expand_as` and `clone`

File: tools/autograd/templates/Functions.cpp [hunk 4]
```diff
@@ -1510,6 +1510,41 @@ Tensor svd_backward(const std::vector<torch::autograd::Variable> &grads, const T
...
+    auto gv = grads[1];
+    
+    auto vt = v.t();
+    
+    if (!eigenvectors) {
+        throw std::runtime_error(std::string("cannot compute backward without "
+                                             "computing eigenvectors in forward pass"));
+    }
+    
+    Tensor result = at::zeros_like(self);
```
Comment (SsnL) [inline tools/autograd/templates/Functions.cpp]: nit: you can just do `Tensor result;` here and assign zeros in the else of `if (gv.defined())`

**Review by ssnl — COMMENTED** [review]
sorry for the late review! looks great except for some minor nits!

**Comment by ezyang** [comment]
Hey @sethah, have you had time to handle the PR comments?

**Comment by sethah** [comment]
@SsnL Sorry for the delay, I think my latest commit should cover the feedback you provided. Thanks!

**Review by ssnl — APPROVED** [review]
Thanks!!
